lib/codecs/src/decoding/framing/chunked_gelf.rs

```diff
@@ -0,0 +1,650 @@
+use crate::BytesDecoder;
+
+use super::BoxedFramingError;
+use bytes::{Buf, Bytes, BytesMut};
+use derivative::Derivative;
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+use tokio;
+use tokio::task::JoinHandle;
+use tokio_util::codec::Decoder;
+use tracing::{info, warn};
+use vector_config::configurable_component;
+
+const GELF_MAGIC: [u8; 2] = [0x1e, 0x0f];
+const GELF_MAX_TOTAL_CHUNKS: u8 = 128;
+const DEFAULT_CHUNKS: [Bytes; GELF_MAX_TOTAL_CHUNKS as usize] =
+    [const { Bytes::new() }; GELF_MAX_TOTAL_CHUNKS as usize];
+const DEFAULT_TIMEOUT_MILLIS: u64 = 5000;
+// TODO: ask what would be an appropriate default value for this
+const DEFAULT_PENDING_MESSAGES_LIMIT: usize = 1000;
+
+const fn default_timeout_millis() -> u64 {
+    DEFAULT_TIMEOUT_MILLIS
+}
+
+const fn default_pending_messages_limit() -> usize {
+    DEFAULT_PENDING_MESSAGES_LIMIT
+}
+
+/// Config used to build a `ChunkedGelfDecoder`.
+#[configurable_component]
+#[derive(Debug, Clone, Default, PartialEq, Eq)]
+pub struct ChunkedGelfDecoderConfig {
+    /// Options for the chunked gelf decoder.
+    #[serde(default, skip_serializing_if = "vector_core::serde::is_default")]
+    pub chunked_gelf: ChunkedGelfDecoderOptions,
+}
+
+impl ChunkedGelfDecoderConfig {
+    /// Build the `ChunkedGelfDecoder` from this configuration.
+    pub fn build(&self) -> ChunkedGelfDecoder {
+        ChunkedGelfDecoder::new(
+            self.chunked_gelf.timeout_millis,
+            self.chunked_gelf.pending_messages_limit,
+        )
+    }
+}
+
+/// Options for building a `ChunkedGelfDecoder`.
+#[configurable_component]
+#[derive(Clone, Debug, Derivative, PartialEq, Eq)]
+pub struct ChunkedGelfDecoderOptions {
+    /// The timeout in milliseconds for a message to be fully received. If the timeout is reached, the
+    /// decoder will drop all the received chunks of the uncomplete message and start over.
+    ///  The default value is 5 seconds.
+    #[serde(
+        default = "default_timeout_millis",
+        skip_serializing_if = "vector_core::serde::is_default"
+    )]
+    pub timeout_millis: u64,
+
+    /// The maximum number of pending uncomplete messages. If this limit is reached, the decoder will start
+    /// dropping chunks of new messages. This limit ensures the memory usage of the decoder's state is bounded.
```

Comment:
The bound is per total messages, but there is no per-message memory usage limit. We can theoretically have a 100GB single message and won't be limited by this setting.

As stated before, should we include a per-message limit?

Reply:
I like the idea of having a configurable bound on the number of pending messages.

The chunked encoding is only used for UDP, yes? Shouldn't that provide a defacto bound on size? How can we have a 100 GB message?

Reply:
> The chunked encoding is only used for UDP, yes? 
Yes, it is intended to use it only for UDP and therefore it would be limited by the UDP packets limit of 65KB.
Nevertheless, as the chunked_gelf is a framing method, nothing blocks user to use that method with other types of sources, for example, with `file` sources. Although, it really does not have sense to use that framing method outside of UDP socket sources, and no one will use that in real environments... So maybe it is ok to leave this as it is

Reply:
Given the nature of gelf's messages, as they are just json, I don't think it would be fine to just truncate the input, as the json message would be most likely broken and the GELF deserialization would fail in nearly all cases after truncating.

Should we instead discard the whole message (including previously stored cunks) if an individual chunk reaches its defined limit or the accumulated chunks limit is reached? I don't know if its worth to do this, but I'm open to implement  it if you see that it would be worth

Reply:
Ah, yes, I think we'd want to discard messages that exceed the limit (in the future we can route them to a "dead letter" output). This is consistent with the other framers (see example: https://github.com/vectordotdev/vector/blob/17466c684bc6d3b0356c976e9639f9b05d306769/lib/codecs/src/decoding/framing/character_delimited.rs#L123-L128).

I think it'd be worth it to remove a DOS risk if it isn't too much effort to add.